pallets/migrations/src/tests.rs

```diff
@@ -0,0 +1,267 @@
+				move || {
+					let mut num_name_fn_calls = num_name_fn_calls.lock().unwrap();
+					*num_name_fn_calls += 1;
+					"migration1"
+				},
+				move |_| -> Weight {
+					let mut num_step_fn_calls = num_step_fn_calls.lock().unwrap();
+					*num_step_fn_calls += 1;
+					1u32.into()
+				},
+			);
+		},
+		&mut || {
+			ExtBuilder::default().build().execute_with(|| {
+				// roll forward until upgraded, should happen before block even increments
+				crate::mock::roll_until_upgraded(true);
+
+				assert_eq!(System::block_number(), 1);
+				assert_eq!(
+					*num_name_fn_calls.lock().unwrap(),
+					1,
+					"migration name needed once"
+				);
+				assert_eq!(
+					*num_step_fn_calls.lock().unwrap(),
+					1,
+					"migration step needed once"
+				);
+				let mut expected = vec![
+					Event::RuntimeUpgradeStarted(),
+					Event::MigrationStarted("migration1".into()),
+					Event::MigrationCompleted("migration1".into(), 1u32.into()),
+					Event::RuntimeUpgradeCompleted(100000001u32.into()), // includes reads/writes
+				];
+				assert_eq!(events(), expected);
+
+				// attempt to roll forward again, block should still not increment, and migration
+				// name fn should be called but pallet_migrations should immediately recognize that
+				// no work needs to be done (and not call step)
+				crate::mock::roll_until_upgraded(true);
+
+				assert_eq!(System::block_number(), 1);
+				assert_eq!(
+					*num_name_fn_calls.lock().unwrap(),
+					2,
+					"migration name needed twice"
+				);
+				assert_eq!(
+					*num_step_fn_calls.lock().unwrap(),
+					1,
+					"migration step not needed again"
+				);
+				expected.append(&mut vec![
+					Event::RuntimeUpgradeStarted(),
+					Event::RuntimeUpgradeCompleted(100000000u32.into()),
+				]);
+				assert_eq!(events(), expected);
+
+				// roll forward a few blocks
+				crate::mock::roll_to(3, false);
+				assert_eq!(
+					*num_name_fn_calls.lock().unwrap(),
+					2,
+					"migration name not needed again"
+				);
+				assert_eq!(
+					*num_step_fn_calls.lock().unwrap(),
+					1,
+					"migration step not needed again"
+				);
+				// assert that no new events have been emitted
+				assert_eq!(events(), expected);
+			});
+		},
+	);
+}
+
+#[test]
+fn on_runtime_upgrade_charges_minimum_two_db_writes() {
+	ExtBuilder::default().build().execute_with(|| {
+		let weight = Migrations::on_runtime_upgrade();
+		assert_eq!(weight, RocksDbWeight::get().writes(2));
+	})
+}
+
+#[test]
+fn overweight_migrations_tolerated() {
+	// pallet-migrations currently tolerates a migration going over-weight. not only does it
+	// tolerate it, but it continues on to the next migration even if it's already overweight.
+	//
+	// The logic behind this is that we would rather go over-weight and risk a block taking too long
+	// (which *might* be "catastrophic") than outright prevent migrations from proceeding (which is
+	// certainly "catastrophic").
+	//
+	// maybe_catastrophic > certainly_catastrophic
```

Comment:
```suggestion
	// maybe_catastrophic < certainly_catastrophic
```
is this "`certainly_catastrophic` is worse than `maybe_catastrophic`"?